### PR TITLE
Assets download & caching

### PR DESCRIPTION
**Pull request by omichel**
This PR implements support for assets download and caching.
It addresses partly #1638.
- [x] Download of `ImageTexture.url`.
- [x] Cache of downloaded asserts.
- [x] `Background.*Url`
- [x] `Camera.noiseMaskUrl`
- [x] `Mesh.url`
- [x] `ContactProperties.*Sound`
- [x] `RotationalMotor.sound`
- [x] `LinearMotor.sound`
- [x] Support for assets download from PROTO nodes.
- [x] Example world.
- [x] Support changing the URL from the scene tree.
- [x] Support changing the URL from the supervisor.
- [x] Clean-up.
- [x] Tests.
- [x] Documentation.
- [x] Change log.

`WbApplication.mWorldLoadTimer`, which was causing the world loading dialog appear only after a few seconds, was removed for the following reasons:
1. it was causing problems with the asset download and would have required a redesign of the code to work properly.
2. it was not very useful as most of the worlds are anyhow displaying this dialog.

### CODE REVIEW

File: src/webots/nodes/WbImageTexture.cpp [hunk 1]
```diff
@@ -383,8 +430,12 @@ void WbImageTexture::pickColor(WbRgb &pickedColor, const WbVector2 &uv) {
   // printf("pickedColor (u=%f, v=%f): (r=%f g=%f b=%f)\n", u, v, pickedColor.red(), pickedColor.green(), pickedColor.blue());
 }
 
-QString WbImageTexture::path() {
-  return WbUrl::computePath(this, "url", mUrl, 0);
+const QString WbImageTexture::path(bool warning) const {
+  if (mUrl->size() == 0)
+    return "";
+  if (mUrl->item(0).startsWith("https://"))
```
Comment (lukicdarkoo) [inline src/webots/nodes/WbImageTexture.cpp:436]: `http://` should be included as well?

**Review by lukicdarkoo — APPROVED** [review]
Nice! I briefly checked and it works fine on Ubuntu

**Comment by omichel** [comment]
I applied your suggestions and re-passed the tests.

File: projects/samples/howto/url/controllers/url/Makefile [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Copyright 1996-2021 Cyberbotics Ltd.
```
Comment (lukicdarkoo) [inline projects/samples/howto/url/controllers/url/Makefile:1]: This file should be included to `projects/samples/howto/Makefile` and the description is missing in `samples-howto.md`.

**Review by lukicdarkoo — CHANGES_REQUESTED** [review]
`hall.wbt` is all black, but I am not sure whether it is related to this PR or something else.

**Comment by omichel** [comment]
Yes, it seems the problem with `hall.wbt` is caused by this PR. I am investigating it.

**Comment by omichel** [comment]
It turns out the `hall.wbt` world has no light at all, so it seems normal that it should appear black.
However, it doesn't appear black on the `develop` branch, and I don't understand why...

**Comment by DavidMansolino** [comment]
> However, it doesn't appear black on the develop branch, and I don't understand why...

It is because the background is producing some light. This is a commun technic used to optimize performance to use only the background to illuminate the scene (and additionally it it looks more realistic in indoors environment where the source of the light is extremely diffuse).

The fact that it doesn't work in this branch probably indicates that the `*IrradianceUrl` images of the background are not loaded.

**Comment by omichel** [comment]
> It is because the background is producing some light. This is a commun technic used to optimize performance to use only the background to illuminate the scene (and additionally it it looks more realistic in indoors environment where the source of the light is extremely diffuse).
>
> The fact that it doesn't work in this branch probably indicates that the *IrradianceUrl images of the background are not loaded.

That makes sense. Thank you!

**Comment by omichel** [comment]
Now the `hall.wbt` is rendered properly (thanks to @DavidMansolino!)

**Review by lukicdarkoo — DISMISSED** [review]
Looks good to me. However, it would be good that @stefaniapedrazzi check it again.

File: src/webots/nodes/WbBackground.cpp [hunk 3]
```diff
@@ -287,199 +389,248 @@ void WbBackground::applyColourToWren(const WbRgb &color) {
...
+bool WbBackground::loadTexture(int i) {
+  if (mTexture[i])
+    return true;
+  QString url;
+  QIODevice *device;
+
+  if (mDownloader[i]) {
+    url = mUrlFields[i]->item(0);
+    if (!mDownloader[i]->error().isEmpty()) {
+      warn(tr("Cannot not retrieve '%1': %2").arg(url).arg(mDownloader[i]->error()));
```
Comment (stefaniapedrazzi) [inline src/webots/nodes/WbBackground.cpp:401]:
```suggestion
      warn(tr("Cannot retrieve '%1': %2").arg(url).arg(mDownloader[i]->error()));
```

File: src/webots/nodes/WbBackground.cpp [hunk 4]
```diff
@@ -287,199 +389,248 @@ void WbBackground::applyColourToWren(const WbRgb &color) {
...
+      return false;
+    url = WbUrl::computePath(this, QString("%1Url").arg(gDirections[i]), mUrlFields[i]->item(0), false);
+    if (url == WbUrl::missingTexture()) {
+      warn(tr("Texture not found: '%1'").arg(mUrlFields[i]->item(0)));
+      return false;
+    }
+    device = new QFile(url);
+    if (!device->open(QIODevice::ReadOnly)) {
+      warn(tr("Cannot open texture file: '%1'").arg(url));
+      return false;
```
Comment (stefaniapedrazzi) [inline src/webots/nodes/WbBackground.cpp:419]: Is it not needed to delete `device` before returning?
Same for the other return cases later in this function.

File: src/webots/nodes/WbBackground.cpp [hunk 5]
```diff
@@ -287,199 +389,248 @@ void WbBackground::applyColourToWren(const WbRgb &color) {
...
+      WbUrl::computePath(this, QString("%1IrradianceUrl").arg(gDirections[i]), mIrradianceUrlFields[j]->item(0), false);
+    if (url.isEmpty()) {
+      warn(tr("%1IrradianceUrl not found: '%2'").arg(gDirections[i], mUrlFields[i]->item(0)));
+      return false;
+    }
+    device = new QFile(url);
+    shouldDelete = true;
+    if (!device->open(QIODevice::ReadOnly)) {
+      warn(tr("Cannot open HDR texture file: '%1'").arg(mIrradianceUrlFields[j]->item(0)));
+      return false;
```
Comment (stefaniapedrazzi) [inline src/webots/nodes/WbBackground.cpp:509]: Is it not needed to delete `device` before returning?

**Review by stefaniapedrazzi — CHANGES_REQUESTED** [review]
Globally everything is fine, but I found a couple of issues:
* as described more in detail in one of the comments, deleting the `ImageTexture` url item (`PBRAppearance` node) doesn't trigger the object appearance change in the 3D view
* `Background`:
    * also in this case deleting one `*Url` value doesn't trigger the scene update (I have to save and revert the world to see the black background)
    * the console message warning about the incomplete cubemap when one of the `*Url` field is empty is no longer printed in the console. I would keep it as it helps users understanding why the background is still black after defining some of the fields

File: src/webots/app/WbApplication.cpp [hunk 6]
```diff
@@ -199,7 +198,7 @@ void WbApplication::linkLibraries(QString projectLibrariesPath) {
 }
 
 void WbApplication::setWorldLoadingProgress(const int progress) {
-  if (!mWorldLoadingProgressDialogCreated && mWorldLoadTimer && (mWorldLoadTimer->elapsed() / 1000) > 2) {
```
[the remark this answers is not in the document]
Comment (omichel) [reply src/webots/app/WbApplication.cpp:202]: I just explained in the PR description.

**Comment by omichel** [comment]
Thank you @stefaniapedrazzi for the in depth review. I believe I fixed all the problems you reported.

**Review by stefaniapedrazzi — CHANGES_REQUESTED** [review]
Thank you! Yes, all my previous comments were fixed.

There is still an issue with the irradiance cubemap.
If the cubemap is complete but the irradiance cubemap is not complete,  Webots crashes with the following log:
```
webots-bin: TextureCubeMap.cpp:73: virtual void wren::TextureCubeMap::prepareGl(): Assertion `mDatas[i]' failed.
```

I can reproduce it by converting to base nodes the `TexturedBackround` PROTO, deleting one of the `*IrradianceUrl` item, saving and reverting the world.
Maybe there are also some update issues, because otherwise it should crash immediately when the irradiance url is delete. Also, previously the following message was printed in the console, but now nothing is printed when deleting the irradiance url
```
WARNING: Background: Incomplete irradiance cubemap
```

**Comment by omichel** [comment]
I cannot reproduce this crash. I tried both on Windows and Linux. Which world did you open to make it crash?

**Comment by stefaniapedrazzi** [comment]
I tested different worlds and I can reproduce it with all of them.
For example the `battery.wbt` world.

It is an assertion failure so I think it is not reproducible if Webots is compiled in release mode.

**Comment by omichel** [comment]
Oh, I see, Wren should be compiled in debug mode...

**Comment by omichel** [comment]
OK, I fixed the crash.

**Comment by stefaniapedrazzi** [comment]
I confirm the crash is gone and now the message is correctly printed in the console.

Could you just check one last behavior?
If I load a modified `battery.wbt` world where the `frontIrradianceUrl` is empty, and I copy the `bottomIrradianceUrl` to the `frontIrradianceUrl`, the rectangle arena walls become darker. But if I remove again the `frontIrradianceUrl` value they doesn't return hell. It seems like the update is not applied.

**Comment by omichel** [comment]
> If I load a modified battery.wbt world where the frontIrradianceUrl is empty, and I copy the bottomIrradianceUrl to the frontIrradianceUrl, the rectangle arena walls become darker. But if I remove again the frontIrradianceUrl value they doesn't return hell. It seems like the update is not applied.

Well spotted! I believe I fixed this one as well.

**Comment by stefaniapedrazzi** [comment]
Thank you!
Now the update is applied correctly and I can clearly see the difference when re-adding an irradiance url (and I get the same after save and revert). Just one last note, the behavior in this PR is now different from before and if the irradiance cubemap is not complete, also the background "baseColor" cubemap is not displayed.

**Comment by omichel** [comment]
> Just one last note, the behavior in this PR is now different from before and if the irradiance cubemap is not complete, also the background "baseColor" cubemap is not displayed.

Yes, that is true, but I believe it's better that way, as we don't want to encourage incomplete maps. Users should fix all the warnings if they want the background to be displayed.

**Review by stefaniapedrazzi — APPROVED** [review]
If the behavior is wanted then everything is fine for me.

**Comment by DavidMansolino** [comment]
> > Just one last note, the behavior in this PR is now different from before and if the irradiance cubemap is not complete, also the background "baseColor" cubemap is not displayed.
> 
> Yes, that is true, but I believe it's better that way, as we don't want to encourage incomplete maps. Users should fix all the warnings if they want the background to be displayed.

Note that adding the constraint to always having a complete irradiance map to display the background is probably fine, but adding the opposite constraint (complete background (color) map for the irradiance to work) is IMO a regression as it is often useful to have the reflection of the background without actually seeing it (e.g. when you are inside a building, by the way, several of the world files use this technic currently).

In any case, the ``TexturedBackground`` PROTO currently doesn't respect the new constraint introduced in this PR (https://github.com/cyberbotics/webots/blob/develop/projects/objects/backgrounds/protos/TexturedBackground.proto#L36-L37)

**Comment by omichel** [comment]
> Note that adding the constraint to always having a complete irradiance map to display the background is probably fine, but adding the opposite constraint (complete background (color) map for the irradiance to work) is IMO a regression as it is often useful to have the reflection of the background without actually seeing it (e.g. when you are inside a building, by the way, several of the world files use this technic currently).
>
> In any case, the TexturedBackground PROTO currently doesn't respect the new constraint introduced in this PR (https://github.com/cyberbotics/webots/blob/develop/projects/objects/backgrounds/protos/TexturedBackground.proto#L36-L37)

You are right, it seems the [hall.wbt](https://github.com/cyberbotics/webots/blob/develop/projects/samples/environments/factory/worlds/hall.wbt) world is broken again for this reason. I could fix it previously as it was reported broken by @lukicdarkoo, but it seems I broke my own fix afterward 😒... I will further investigate and fix this.

**Comment by omichel** [comment]
That should be fixed in #2695: the color map should be either complete or empty (e.g., 6 or 0 textures loaded) so that the irradiance map is loaded.

**Pull request merged**